### DISCUSSION TITLE
Suggestion of the package organization

### DISCUSSION THREAD

**Discussion by jAniceto** [wigging/chemics#22] [Ideas]
Hey.

This might be a bit premature but I'll leave a suggestion on the organization of the package. As the package starts to cover areas of Chemical Engineering other than fluidization/reaction (I am currently focusing in adsorption) it might be useful to create sub-packages. Here is a suggestion:

```
chemics
    core
        atomic_elements
        dimensionless_numbers
        etc
    fluidization
        bed_expansion_factor
        bubble_fraction
        etc
    adsorption
        mass_transfer_correlations
        isotherms
        etc
```

Also, I'm anticipating some concepts can appear in slightly different forms in different areas of Chem. Eng.. Having them side-by-side or "unorganized" might cause confusion. Just an example: The Reynolds number can often appear with porosity in packed bed problems. However adding a second Reynolds to the same module might be confusing. 

I understand this might be a major change so this is mostly just food for though at the moment.

**Comment by wigging** [comment]
I completely agree with you that the package needs to be better organized. Since I have been the main contributor to Chemics, I haven't worried about the structure of the package. My main priority has been to keep adding to it as I have time. But yes, it needs to be better organized to encourage others to contribute and to prevent confusion.

Two ideas come to mind on how to organize the package's components. The first is to group the modules (files) into sub packages which are related to a specific type of reactor. For example, functions and modules that relate to a bubbling fluidized bed reactor would be in a bfb_reactor sub package or functions/modules for a fixed bed reactor would be in a fixed_bed sub package. The second idea is to group functions and modules according to their specific area of chemical engineering which is what you have suggested. This approach would probably be more suitable and prevent feature overlap.

Currently, I don't think it would be much work to reorganize as long as we agree on a package structure and don't make any more contributions until after everything is refactored. But this should be done sooner rather than later because it will be more work as the package gets bigger.

**Reply by jAniceto** [reply]
Good to hear. Regarding the organization strategy my suggestion was just that, a suggestion. I am not sure on the best way to organize. 

Your first idea could also work well but seems to result in a very fine subdivision, leading to many sub-packages. Another option could be to go with a coarse subdivision like `core`, `reaction`, `separation`, `thermodynamics`, etc. This would result in less sub-packages, I think. My initial suggestion might be a middle ground.

**Reply by wigging** [reply]
Keep in mind that a Python file is a module and a package is a collection of modules. Sub-packages may not be necessary if everything can be done with a module. Ultimately the contents of the Chemics package will determine how things are organized. So the organization of the package will likely change as Chemics obtains new functionality.

**Comment by wigging** [comment]
Here's my attempt at a better file structure for the Chemics package. The names of the modules in the sub-packages could be simplified as part of the reorganization.

```
chemics
|-- data
|    |-- file.csv
|    |-- file.csv
|    |-- file.txt
|    |-- etc.
|-- chemical
|    |-- atomic_elements.py
|    |-- proximate_bases.py
|    |-- etc.
|-- fluidization
|    |-- bed_expansion_factor.py
|    |-- choking_velocity.py
|    |-- transport_velocity.py
|    |-- etc.
|-- gas_properties
|    |-- gas_density.py
|    |-- gas_pressure.py
|    |-- etc.
|-- pyrolysis
|    |-- devol_time.py
|    |-- etc.
|-- solid_properties
|    |-- wood_heat_capacity.py
|    |-- wood_thermal_conductivity.py
|    |-- etc.
|-- conversions.py
|-- dimensionless_numbers.py
```

**Reply by jAniceto** [reply]
This seems good. It allows the package to grow without needing to do much forward thinking now.
